### PR TITLE
fix(index): discard stale records

### LINKED ISSUE #341: Discard index results that have a conflicting source hash
We notice duplicate entries in our index (same path different source hash), and looking at the timeline, this is how they occur:
1. a file is replaced in SharePoint by another
1. the index gets a deleted event for the old entry, so it deletes the row
1. the index gets a modified event for the new entry, it fetches the HTML, but gets the **old** entry, so it adds back the evicted item
1. the index gets another modified event for the new entry, it again fetches the HTML and gets the **new** entry, so it adds a second row with the same path

### PR DESCRIPTION
**Pull request by dominique-pfister**
fix #341

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -153,6 +153,34 @@ async function handleDelete({ config, handler }, change, log) {
...
+  if (Number.isNaN(lastModifiedMs)) {
+    // last modified date is unusable
+    return false;
+  }
+  const eventTimeMs = Date.parse(change.time);
+  if (Number.isNaN(eventTimeMs)) {
+    // event time is unusable
+    return false;
+  }
+  return eventTimeMs > lastModifiedMs;
```
Comment (tripodsan) [inline src/index.js:181]: so, if the event is newer than the last modifed, it is outdated ???

Comment (dominique-pfister) [reply src/index.js:181]: Not quite: if the UIDs (or source hashes) are different, and the UID at the event time is different than the UID of the HTML found, and the HTML is older, **then** it is outdated. I'll add that to the documentation of the function.

**Review by tripodsan — APPROVED** [review]
not sure about what exactly is outdated.... maybe improve the js doc.

**Comment by dominique-pfister** [comment]
> not sure about what exactly is outdated.... maybe improve the js doc.

Thanks, I updated it as follows:
```
 * Return a flag indicating whether a index record contains outdated data. This
 * is true if:
 *
 * - the change UID and the HTML UID (or source hash) are different
 * - the change event time is later than the last modified of the HTML
 *
 * In that case, the change reported relates to a different and more recent
 * item, so we shouldn't add an index record for an outdated item.
 *
 * @returns true if the record is outdated and shouldn't be used for indexing
```

**Pull request merged**